### ISSUE TITLE
Incorrect blueprint index generation

### ISSUE THREAD

**Issue by kennyweiss** [LLNL/axom#23]
In sidre and spio, there is a function to automatically generate the "blueprint index" from a "mesh blueprint"-conforming group in the sidre hierarchy.  This is used by Visit's mesh blueprint plugin, to visualize mesh data output by sidre. 

In the current implementation, the function takes the name of the mesh as one of its parameters ("mesh_name"), but for hdf5 files, it should really be using the path to the mesh. 

There might need to be some logic in the function's implementation to correct the provided path for hdf5 files.  (Possibly because hdf5 renames the top-level group?)

See: 
* https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html
* https://github.com/LLNL/axom/blob/b8d1ae4cdfd6698c428025b362de2faa463b7501/src/axom/sidre/core/DataStore.hpp#L375-L400
* https://github.com/LLNL/axom/blob/b8d1ae4cdfd6698c428025b362de2faa463b7501/src/axom/sidre/spio/IOManager.hpp#L165-L187

**Comment by cyrush** [comment]
@kennyweiss for Sidre meshes, this should be the path in Sidre to the root of the mesh.

**Comment by nselliott** [comment]
I recall that I've produced a blueprint using the SPIO interface that is readable by VisIt.  I'll try to dig that up.

**Comment by kennyweiss** [comment]
Thanks @cyrush  and @nselliott.

I was planning to make the examples a bit more complicated as well.  E.g. where the mesh is not located at the root of the hierarchy.

**Comment by nselliott** [comment]
@kennyweiss I was mistaken--the example I have was from calling conduit's generate_index directly, not using sidre at all.  Nonetheless it is a working example of using this kind of feature to generate VisIt-readable files, so we should get SPIO to produce effectively the same thing if it can't already do so.

**Comment by nselliott** [comment]
@kennyweiss and @cyrush I just gave you on our network a tarball of a simple blueprint-based json vis dump produced by SAMRAI calling conduit's generate_index.  It contains only one field with a linear profile on a single-level mesh.  If SPIO can't do this, then we need to fix it, which is probably a matter of getting the names and string arguments correct.

**Comment by kennyweiss** [comment]
Thanks @nselliott -- I was able to open that dataset in visit.

The problem might be restricted to cases where the mesh isn't located at the root of the hierarchy.

**Comment by kennyweiss** [comment]
@nselliott -- We also need to support the case where we are adding a mesh to an existing blueprint index group.

I added the following change to ``sidre/DataStore.cpp`` support this:

```diff
-  Group* bpindex = getRoot()->createGroup(index_path);
+  Group* bpindex = getRoot()->hasGroup(index_path)
+      ? getRoot()->getGroup(index_path)
+      : getRoot()->createGroup(index_path);
```

**Comment by nselliott** [comment]
I've been looking at the SPIO-generated blueprint index in the sidre_createdatastore_ex example, and it doesn't appear that there is anything wrong with the blueprint index there.  What doesn't work is that VisIt is unable to generate a plot when the data for explicit coords are in external arrays.  This might be a VisIt bug @cyrush ?  If I hack sidre_createdatastore_ex to make the coords data be owned by the coords Views, then VisIt can plot the tiny mesh created in that example.  VisIt can handle the external arrays that hold the field data but not the coords.

I still need to investigate what happens if the mesh isn't at the root of the hierarchy.

**Comment by cyrush** [comment]
@nselliott I can take a look, if you have a file set that already created it that triggers the issue can you give it to me , that would help me attack it faster.

**Comment by nselliott** [comment]
@cyrush I just gave you two datasets.  bpspio.tar.gz comes straight from running sidre_createdatastore_ex compiled in develop and it doesn't work.  bpspio2.tar.gz is changed to have the explicit coords in the Views' internal data instead of external arrays, and it works.

**Comment by nselliott** [comment]
I have a fix coming to make IOManager's method work with a path to the mesh anywhere in the hierarchy.  The problem was that the "mesh_name" argument was used both for the name of the mesh held by the blueprint index and for the mesh in the data files, when there is no need for those names to be the same, so there should be another argument.  I'll change "mesh_name" to "mesh_path" so that it can find the mesh anywhere, and add a "blueprint_name" argument that should be a simple string.

I need to regenerate shroud code to make this all ready.

**Comment by kennyweiss** [comment]
Thanks @nselliott.

So, what will the new signature be?

It sounds like we need:

1. The path to the root of the blueprint index  (currently `index_path`)
2. The path the the root of the current mesh blueprint (currently `domain_path` ? )
3. The name of the mesh (currently `mesh_name` )
4. The path to the root of the mesh (proposed `mesh_path`)  ?

Can we use 2. and 3.  (and our knowledge of the mesh blueprint) to compose the "mesh_path" ("4" in the above list) without adding a new parameter that users need to worry about?  Or is an explicit path necessary in some cases?

**Comment by nselliott** [comment]
I have it currently:

```
IOManager:: writeBlueprintIndexToRootFile(DataStore* datastore,
                                               const std::string& domain_path,
                                               const std::string& file_name,
                                               const std::string& mesh_path,
                                               const std::string& blueprint_name);
```

datastore -- The DataStore holding all this data
domain_path -- Sidre path from root of datastore to the mesh blueprint
file_name -- Name of root file created by IOManager::write()
mesh_path -- HDF5 path from datagroup_%07d inside file created by IOManager::write() to the mesh blueprint
blueprint_name -- Name of the mesh to be stored in the root file at /blueprint_index/_blueprint_name_

This has to be called after IOManager::write().  If write() was called to write the entire datastore->getRoot(), then domain_path and mesh_path should be the same, but there's no requirement for write() to be called in that manner.  mesh_path is going to always be somewhat redundant, because it has to be a subset of domain_path, but I don't know if it's immediately obvious which subset it is.

**Comment by nselliott** [comment]
I need to think about this some more.  This is not a very intuitive interface.

**Comment by nselliott** [comment]
One alternative is not to add the new blueprint_name argument, and use the final name in the mesh_path for that purpose.  So if the mesh_path is "foo/bar/domains/domain_0/mesh", the index will be added to the rootfile at "blueprint_index/mesh".  This would avoid changing the API but still enables users to put real paths in the mesh_path argument.

**Comment by cyrush** [comment]
@nselliott  I have a gap here, where is `sidre_createdatastore_ex`? 

`examples/sidre_createdatastore` doesn't seem right. 

Searching the repo only yields a reference in the codes: src/axom/sidre/docs/sphinx/figs/README.md, but I don't see the code example?

**Comment by nselliott** [comment]
@cyrush The code is at src/axom/sidre/examples/sidre_createdatastore.cpp .  Executable at examples/sidre_createdatastore_ex when you build everything.

**Comment by cyrush** [comment]
@nselliott I was tripped up by external, since the x,y,z are actually defined on a sidre buffer. 
regardless: i found the issue in the visit plugin, i'll get it fixed for VisIt 3.0.1.
Thanks for giving me the files!

**Comment by nselliott** [comment]
Closing because of merge of pull request #51
